File: class_primer/class_primer__fighting_game.cs
using System.Linq;
using static System.Console;

internal class Program
{
    class Player
    {
        internal int Hlt { get; set; }
        internal int[] Atk { get; set; }
        internal int[] Frm { get; set; }

        internal Player( int[] a )
        {
            Hlt = a[0];
            Frm = new int[] { a[1], a[3], a[5] };
            Atk = new int[] { a[2], a[4], a[6] };
        }

        internal void Enhance( int a )
        {
            foreach( int i in Enumerable.Range( 0, 3 ).Where( i => i != a ) ) {
                Atk[i] += 5;
                Frm[i] = System.Math.Max( Frm[i] - 3, 1 );
            }
            return;
        }
    }

    static void Main()
    {
        int[] n = ReadLine().Split().Select( int.Parse ).ToArray();
        Player[] a = new Player[n[0]];

        foreach( int i in Enumerable.Range( 0, n[0] ) ) {
            a[i] = new Player( ReadLine().Split().Select( int.Parse ).ToArray() );
        }
        foreach( int _ in Enumerable.Range( 0, n[1] ) ) {
            int[] t = ReadLine().Split().Select( x => int.Parse( x ) - 1 ).ToArray();
            int a1 = t[1];
            int a2 = t[3];
            Player p1 = a[t[0]];
            Player p2 = a[t[2]];

            if( p1.Hlt <= 0 || p2.Hlt <= 0 ) {
                continue;
            }
            if( p1.Atk[a1] == 0 && p1.Frm[a1] == 0 ) {
                p1.Enhance( a1 );
                if( p2.Atk[a2] == 0 && p2.Frm[a2] == 0 ) {
                    p2.Enhance( a2 );
                } else {
                    p1.Hlt -= p2.Atk[a2];
                }
            } else if( p2.Atk[a2] == 0 && p2.Frm[a2] == 0 ) {
                p2.Hlt -= p1.Atk[a1];
                p2.Enhance( a2 );
            } else {
                if( p1.Frm[a1] < p2.Frm[a2] ) {
                    p2.Hlt -= p1.Atk[a1];
                } else if( p2.Frm[a2] < p1.Frm[a1] ) {
                    p1.Hlt -= p2.Atk[a2];
                }
            }
        }
        WriteLine( a.Count( x => 0 < x.Hlt ) );
        return;
    }
}

File: class_primer/class_primer__fighting_game.py
class player:
    def __init__(self, a):
        self.a = [a[2], a[4], a[6]]
        self.f = [a[1], a[3], a[5]]
        self.hp = a[0]

    def enhance(self, t):
        for i in range(3):
            if i != t:
                self.a[i] += 5
                self.f[i] = max(self.f[i] - 3, 1)

n, k = list(map(int, input().split()))
o = 0
l = []

for _ in range(n):
    l.append(player(list(map(int, input().split()))))
for _ in range(k):
    p1, t1, p2, t2 = list(map(int, input().split()))

    p1 -= 1
    t1 -= 1
    p2 -= 1
    t2 -= 1
    if l[p1].hp <= 0 or l[p2].hp <= 0:
        continue
    if l[p1].a[t1] == 0 and l[p1].f[t1] == 0:
        l[p1].enhance(t1)
        if l[p2].a[t2] == 0 and l[p2].f[t2] == 0:
            l[p2].enhance(t2)
        else:
            l[p1].hp -= l[p2].a[t2]
    elif l[p2].a[t2] == 0 and l[p2].f[t2] == 0:
        l[p2].hp -= l[p1].a[t1]
        l[p2].enhance(t2)
    else:
        if l[p1].f[t1] < l[p2].f[t2]:
            l[p2].hp -= l[p1].a[t1]
        elif l[p2].f[t2] < l[p1].f[t1]:
            l[p1].hp -= l[p2].a[t2]
for i in l:
    if 0 < i.hp:
        o += 1
print(o)
